src/aruco_pose/src/aruco_pose.cpp

```diff
@@ -0,0 +1,218 @@
+
+  // Configure the median filter. 6 refers to the number of channels in the multi-channel filter
+  // Note: it is necessary to have/declare a parameter named 'number_of_observations'
+  // in the parameter server.
+  median_filter_->configure(
+    6, "", "number_of_observations",
+    this->get_node_logging_interface(), this->get_node_parameters_interface());
+
+  RCLCPP_INFO(LOGGER, "Pose estimator node started!");
+}
+
+void ArucoPose::image_raw_callback(
+  const sensor_msgs::msg::Image::ConstSharedPtr & rgb_msg)
+{
+  // Convert ROS image message to cv::Mat
+  cv_bridge::CvImagePtr cv_ptr_rgb =
+    cv_bridge::toCvCopy(rgb_msg, sensor_msgs::image_encodings::BGR8);
+
+  // Detect the markers from the incoming image
+  cv::aruco::detectMarkers(
+    cv_ptr_rgb->image, dictionary_, markerCorners_, markerIds_, parameters_,
+    rejectedCandidates_);
+
+  // Exclude instances where no markers are detected
+  try {
+    if (!markerIds_.empty()) {
+      // rvecs: rotational vector
+      // tvecs: translation vector
+      std::vector<cv::Vec3d> rvecs, tvecs;
+
+      // Pose estimation happens here
+      cv::aruco::estimatePoseSingleMarkers(
+        markerCorners_, physical_marker_size_, cameraMatrix_, distCoeffs_, rvecs, tvecs);
+
+      cv::Mat R;
+      //  Convert the rvecs to a rotation matrix
+      for (size_t i = 0; i < rvecs.size(); ++i) {
+        cv::Rodrigues(rvecs[i], R);
+      }
+
+      // Access each element of the R matrix for easy rpy calculations
+      double r11 = R.at<double>(0, 0), r21 = R.at<double>(1, 0), r31 = R.at<double>(2, 0),
+        r32 = R.at<double>(2, 1), r33 = R.at<double>(2, 2);
+
+      double roll, pitch, yaw;
+
+      // rpy calculation
+      roll = std::atan2(r32, r33);
+      pitch = std::asin(-r31);
+      yaw = std::atan2(r21, r11);
+
+      auto tranlsation = tvecs[0];
+
+      // Construct the raw rpy_xyz of the marker
+      std::vector<double> raw_rpyxyz =
+      {roll, pitch, yaw, tranlsation[0], tranlsation[1], tranlsation[2]};
+
+      // Median filter gets applied
+      median_filter_->update(raw_rpyxyz, median_filtered_rpyxyz);
+
+      // Add to the tf frame here for the sample
+      geometry_msgs::msg::TransformStamped transformStamped_tag;
+
+      transformStamped_tag.header.stamp = this->now();
+      transformStamped_tag.header.frame_id = this->get_parameter("camera_tf_frame").as_string();
+      transformStamped_tag.child_frame_id = this->get_parameter("sample_name").as_string();
+
+      transformStamped_tag.transform.translation.x = median_filtered_rpyxyz[3] +
+        this->get_parameter("offset_on_marker_x").as_double();
+      transformStamped_tag.transform.translation.y = median_filtered_rpyxyz[4] +
+        this->get_parameter("offset_on_marker_y").as_double();
+      transformStamped_tag.transform.translation.z = median_filtered_rpyxyz[5];
+      transformStamped_tag.transform.rotation = toQuaternion(roll, pitch, yaw);
```

Comment:
This should be the median filtered rotation.

Reply:
roll pitch yaw is changed to median filter values:
https://github.com/maffettone/erobs/blob/5c32dc6fe6c0a002d6ecb2d3e5b09daddd49ae45/src/aruco_pose/src/aruco_pose.cpp#L166-L170